### PR TITLE
New Keypoint Heads and Losses

### PR DESCRIPTION
**Pull request by JSabadin**
- Added sigma and area factor to the metrics.

- The implicit keypoint bbox head now uses a new OKC loss instead of the detection loss.

- Added a new Efficient keypoint bbox head and loss.

- Updates to README:
  - Instructions for `luxonis-train inspect` and `evaluate` commands.
  - Description of the new head and loss.
  - Updated weight names for the implicit keypoint bbox loss.

### CODE REVIEW

File: luxonis_train/utils/assigners/tal_assigner.py [hunk 1]
```diff
@@ -121,7 +121,13 @@ def forward(
 
         out_mask_positive = mask_pos_sum.bool()
 
-        return assigned_labels, assigned_bboxes, assigned_scores, out_mask_positive
+        return (
```
Comment (kozlov721) [inline luxonis_train/utils/assigners/tal_assigner.py:124]: The return type of the function has to be changed.

File: luxonis_train/utils/assigners/tal_assigner.py [hunk 2]
```diff
@@ -121,7 +121,13 @@ def forward(
 
         out_mask_positive = mask_pos_sum.bool()
 
-        return assigned_labels, assigned_bboxes, assigned_scores, out_mask_positive
```
Comment (kozlov721) [inline luxonis_train/utils/assigners/tal_assigner.py:124]: There are two `return`s in `forward`. Only one is updated to return 5 items.

File: luxonis_train/utils/assigners/atts_assigner.py [hunk 3]
```diff
@@ -124,6 +124,7 @@ def forward(
             assigned_bboxes,
```
Comment (kozlov721) [inline luxonis_train/utils/assigners/atts_assigner.py:124]: Same as `tal_assigner`.

File: luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+        self.varifocal_loss = VarifocalLoss()
+        self.class_bbox_loss_weight = class_bbox_loss_weight
+        self.iou_loss_weight = iou_loss_weight
+        self.regr_kpts_loss_weight = regr_kpts_loss_weight
+        self.vis_kpts_loss_weight = vis_kpts_loss_weight
+
+    def prepare(
+        self, outputs: Packet[Tensor], labels: Labels
+    ) -> tuple[Tensor, Tensor, Tensor, Tensor, Tensor, Tensor]:
```
Comment (kozlov721) [inline luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py:133]: Incorrect return type annotation.

File: luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+            "iou": loss_iou.detach(),
+            "regression": regression_loss.detach(),
+            "visibility": visibility_loss.detach(),
+        }
+
+        return loss, sub_losses
+
+    def _preprocess_bbox_target(
+        self, bbox_target: Tensor, batch_size: int, scale_tensor: Tensor
+    ):
```
Comment (kozlov721) [inline luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py:312]: Missing return type

File: luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+        for id, count in zip(sample_ids, counts):
+            out_target[id, :count] = bbox_target[bbox_target[:, 0] == id][:, 1:]
+
+        scaled_target = out_target[:, :, 1:5] * scale_tensor
+        out_target[..., 1:] = box_convert(scaled_target, "xywh", "xyxy")
+        return out_target
+
+    def _preprocess_kpts_target(
+        self, kpts_target: Tensor, batch_size: int, scale_tensor: Tensor
+    ):
```
Comment (kozlov721) [inline luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py:331]: Missing return type

File: luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+            keypoints_i = kpts_target[kpts_target[:, 0] == i]
+            scaled_keypoints_i = keypoints_i[:, 2:].clone()
+            batched_keypoints[i, : keypoints_i.shape[0]] = scaled_keypoints_i.view(
+                -1, self.n_kps, 3
+            )
+            batched_keypoints[i, :, :, :2] *= scale_tensor[:2]
+
+        return batched_keypoints
+
+    def kpts_decode(self, anchor_points: Tensor, pred_kpts: Tensor):
```
Comment (kozlov721) [inline luxonis_train/attached_modules/losses/efficient_keypoint_bbox_loss.py:350]: Missing return type

File: luxonis_train/nodes/efficient_keypoint_bbox_head.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        anchor_points_y = anchor_points_transposed[1].view(1, -1, 1)
+
+        y[:, :, 0::3] = (y[:, :, 0::3] * 2.0 + (anchor_points_x - 0.5)) * stride_tensor
+        y[:, :, 1::3] = (y[:, :, 1::3] * 2.0 + (anchor_points_y - 0.5)) * stride_tensor
+        y[:, :, 2::3] = y[:, :, 2::3].sigmoid()
+
+        return y
+
+    def _process_to_bbox_and_kps(
+        self, output: tuple[list[Tensor], Tensor, Tensor]
```
Comment (kozlov721) [inline luxonis_train/nodes/efficient_keypoint_bbox_head.py:177]: `output` has incorrect type

File: luxonis_train/attached_modules/metrics/object_keypoint_similarity.py [hunk 9]
```diff
@@ -156,48 +169,115 @@ def compute(self) -> Tensor:
...
+
 
 def fix_empty_tensors(input_tensor: Tensor) -> Tensor:
     """Empty tensors can cause problems in DDP mode, this methods corrects them."""
     if input_tensor.numel() == 0 and input_tensor.ndim == 1:
         return input_tensor.unsqueeze(0)
     return input_tensor
+
+
+def set_sigmas(
```
Comment (kozlov721) [inline luxonis_train/attached_modules/metrics/object_keypoint_similarity.py:226]: I'd rename it to `compute_sigmas` or `get_sigmas`. `set_sigmas` sounds like it should not return anything.

File: luxonis_train/attached_modules/metrics/object_keypoint_similarity.py [hunk 10]
```diff
@@ -156,48 +169,115 @@ def compute(self) -> Tensor:
...
+            )
+        else:
+            warn_msg = "Default sigma of 0.04 is being used for each keypoint."
+            if class_name:
+                warn_msg = f"[{class_name}] {warn_msg}"
+            logger.warning(warn_msg)
+            return torch.tensor([0.04] * n_keypoints, dtype=torch.float32)
+
+
+def set_area_factor(area_factor: float | None, class_name: str | None) -> float:
```
Comment (kozlov721) [inline luxonis_train/attached_modules/metrics/object_keypoint_similarity.py:274]: Same like with `set_sigmas`

**Review by kozlov721 — COMMENTED** [review]
Left a few minor comments, mostly related to types. Otherwise LGTM!

**Comment by JSabadin** [comment]
I have fixed the type annotations and function renamings as @kozlov721 proposed.

**Comment by kozlov721** [comment]
> I have fixed the type annotations and function renamings as @kozlov721 proposed.

Thanks. Also please update the tests (assigner tests) so the checks pass. They should all pass now; I temporarily disabled the MacOS tests.

**Review by kozlov721 — APPROVED** [review]
LGTM

**Review by conorsim — APPROVED** [review]
I'm lacking domain knowledge but from a quick look it looks good
